### PR TITLE
feat(runner): write "Last Verified On" dates back into source docs

### PR DESCRIPTION
**Pull request by hawkeyexl**
## What

Adds a marker-driven **"Last Verified On"** write-back: authors place a static `verified` marker carrying a test or spec `id`; after a run Doc Detective resolves that id and — **only on PASS** — writes today's date back into the marker. On failure the date is left untouched, so a fresh date means "verified passing" and the badge silently ages.

End-user docs can surface a freshness/trust badge from this, and in CI the GitHub Action's existing `create_pr_on_change` turns the date edits into a PR with **zero Action changes**.

## How it works

- **Author owns the `id`; Doc Detective owns the `date`.** No content-hash guessing, no orphans, no inferred placement.
- **Two marker forms:**
  - Inline comment — markdown/MDX (`{/* … */}`, `<!-- … -->`, `[comment]: #`), AsciiDoc (`// ( … )`), HTML/DITA comments.
  - Page-level YAML front matter — `doc-detective.verified.{id → date}`.
- **Optional `badge`** — maintains a shields.io static badge image (`Last verified …`) on the line after the marker, so it renders with no theme work. DD only writes the URL; shields.io renders it client-side (no network call).
- **Idempotent & form-preserving** — re-runs are byte-identical; an MDX `{/* */}` marker is never rewritten as `<!-- -->` (which would break the MDX build).
- **Unknown id → warn and skip** (catches typos / deleted tests without failing the run).
- **Coverage** — the writer scans the **union** of the report's content paths and the full detected input set, so a prose-only page that produces no spec (referencing a test defined elsewhere) is still updated.

## Scope

- **No config object, no CLI flag, no schema change.** Always-on; acts only where a marker exists.
- Contained in `src/core`: pure helpers + `applyVerifiedMarkers` in `core/utils.ts`, wired into `runTests`; `qualifyFiles` side-channels the detected file set.

## Tests

- `test/verified-markers.test.js` — 21 unit tests (shields.io dash-doubling, MDX form preservation, idempotency, aging, id resolution).
- `test/verified-writeback.test.js` — 13 filesystem-integration tests (all formats, badge, front matter, spec-id vs test-id, PASS/FAIL/SKIPPED, unknown-id warning, byte-idempotency) + 2 real `runTests` E2Es (incl. the cross-file prose-only case).
- **34 passing**; clean `tsc` compile.

These are deliberately **not** committed `test/core-artifacts/*.spec.json` fixtures: the feature mutates source files, so a committed fixture would be rewritten with the current date every run and dirty the tree. Rationale is in the ADR.

## Docs impact

User-facing (new marker syntax + badge). Reference/how-to docs for the marker should follow; flagging here rather than silently. Design + decision recorded in `adrs/01046-last-verified-on-marker-writeback.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added automatic “Last Verified On” write-back for passing results, updating inline verified markers and YAML front matter while preserving body content.
  * Added optional Shields.io verification badges with consistent, format-aware rendering and idempotent updates.
  * Supports marker ids mapped to test- or spec-level outcomes, including cross-file and prose-only inputs; unknown ids leave content unchanged with warnings.
* **Documentation**
  * Documented the marker-based write-back behavior in a new ADR.
* **Tests**
  * Added comprehensive unit and integration tests for marker/badge syntaxes, idempotency, and edge cases.

### CODE REVIEW

File: src/core/index.ts [hunk 1]
```diff
@@ -189,6 +193,19 @@ async function runTests(config: any, options: any = {}) {
...
   cleanTemp();
 
+  // "Last Verified On" write-back: update any `verified` markers in the source
+  // docs with today's date where the referenced test/spec passed. No-op when
+  // no markers exist; never throws (warns and continues on per-file errors).
+  try {
+    applyVerifiedMarkers({
+      config,
+      results,
+      files: (resolvedTests && resolvedTests._qualifiedFiles) || [],
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/index.ts]: Fixed in 0007f150 — `resolvedTests` is always truthy here (runTests returns early otherwise), so the `resolvedTests &&` guard was redundant. Simplified to `resolvedTests._qualifiedFiles || []`.

File: src/core/utils.ts [hunk 2]
```diff
@@ -1154,6 +1161,323 @@ function replaceEnvs(stringOrObject: any): any {
...
+      const img = shieldsBadge(format, date);
+      const body = VERIFIED_IMG_BODY[format] || VERIFIED_IMG_BODY.markdown;
+      const imgLineRe = new RegExp(`^([ \\t]*\\r?\\n[ \\t]*)(?:${body})`);
+      const tail = content.slice(marker.end);
+      const mm = tail.match(imgLineRe);
+      if (mm) {
+        content =
+          content.slice(0, marker.end) + mm[1] + img + content.slice(marker.end + mm[0].length);
+      } else {
+        content = content.slice(0, marker.end) + "\n" + img + content.slice(marker.end);
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts]: Fixed in 0007f150 — both the badge insert in `applyVerifiedToContent` and the front-matter writer now detect the file's existing newline (`\r\n` vs `\n`) and reuse it, so CRLF-authored docs stay byte-idempotent. The front-matter writer also switched to a surgical `date:`-line edit (no whole-document re-serialization).

File: src/core/utils.ts [hunk 3]
```diff
@@ -1154,6 +1161,323 @@ function replaceEnvs(stringOrObject: any): any {
...
+      }
+    }
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts:1595]: Fixed in 0007f150 — the badge-insert branch now reuses the file's detected EOL instead of hardcoding `\n`.

File: src/core/utils.ts [hunk 4]
```diff
@@ -1154,6 +1161,323 @@ function replaceEnvs(stringOrObject: any): any {
...
+  doc.setIn(["doc-detective", "verified", "date"], date);
+  const newBlock = doc.toString().replace(/\r?\n$/, "");
+  return (
+    content.slice(0, fm.index!) +
+    "---\n" +
+    newBlock +
+    "\n---" +
+    content.slice(fm.index! + fm[0].length)
+  );
+}
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts:1737]: Fixed in 0007f150 — front-matter writes reuse the detected EOL, and the write is now a surgical edit of only the `date:` line rather than a full YAML re-serialization (which previously reflowed sibling fields).

File: src/core/utils.ts [hunk 5]
```diff
@@ -1154,6 +1161,323 @@ function replaceEnvs(stringOrObject: any): any {
...
+    try {
+      if (!fs.existsSync(file)) continue;
+      const ext = (file.split(".").pop() || "").toLowerCase();
+      const format = VERIFIED_FORMAT_BY_EXT[ext];
+      if (!format) continue;
+
+      const original = fs.readFileSync(file, "utf8");
+      const inline = parseVerifiedMarkers(original, format);
+      const fmId = frontMatterVerifiedId(original);
+      const ids = new Set(inline.map((m) => m.id));
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts:1797]: Done in 0007f150 — added `if (!original.includes("verified")) continue;` before the per-file regex/YAML scans. Both marker forms contain the literal `verified`, so this safely short-circuits files with no markers.

File: src/core/index.ts [hunk 6]
```diff
@@ -189,6 +193,19 @@ async function runTests(config: any, options: any = {}) {
...
+    });
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/index.ts]: Addressed in 0007f150 — on the pre-resolved (DOC_DETECTIVE_API) path `_qualifiedFiles` is unavailable, so we now emit a `debug` log noting prose-only pages aren't scanned, and the ADR Consequences documents the limitation. Report content paths are still updated on that path.

File: test/verified-writeback.test.js [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+import fs from "node:fs";
+import os from "node:os";
+import path from "node:path";
+
+before(async function () {
+  const { expect } = await import("chai");
+  global.expect = expect;
+});
+
+const TODAY = verifiedDate();
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply test/verified-writeback.test.js]: Fixed in 0007f150 — `TODAY` is recomputed in `beforeEach` instead of at module load, closing the UTC-rollover window.

**Comment by hawkeyexl** [comment]
Thanks for the thorough review — addressed in `0007f150`:

- **YAML front-matter re-serialization (Important):** confirmed real — a test with `tags: [api, rest]` showed `doc.toString()` reflowing it to `[ api, rest ]`. Switched `upsertFrontMatterVerified` to a **surgical text edit of only the `date:` line** within the block-style `verified:` mapping (no whole-document re-stringify), so sibling fields keep their exact formatting. Added a sibling-field-stability test plus a front-matter byte-idempotency check. Flow-style front matter is left untouched (returns content unchanged) rather than risk a reflow.
- **EOL preservation:** badge inserts and front-matter writes now reuse the file's detected newline (CRLF/LF).
- **API-path gap:** on the pre-resolved (`DOC_DETECTIVE_API`) path `_qualifiedFiles` is unavailable; we now log a `debug` note that prose-only pages aren't scanned, and the ADR Consequences documents it. Report content paths are still updated there.
- **Nit — `resolveVerifiedId` not used internally:** intentional. `applyVerifiedMarkers` builds an O(1) `resultById` map in a single pass (it resolves many ids per run), whereas `resolveVerifiedId` is the documented single-lookup helper and is unit-tested directly. Kept both; they share the same specId/testId roll-up semantics.
- **Nit — no AsciiDoc/DITA badge tests:** added badge idempotency integration tests for both formats.

File: src/core/utils.ts [hunk 8]
```diff
@@ -1154,6 +1161,366 @@ function replaceEnvs(stringOrObject: any): any {
...
+  let vIndent = "";
+  for (let i = 0; i < lines.length; i++) {
+    const m = lines[i].match(/^([ \t]*)verified:[ \t]*$/);
+    if (m) {
+      vIndex = i;
+      vIndent = m[1];
+      break;
+    }
+  }
+  if (vIndex === -1) return content; // flow style / unusual shape — leave untouched
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts]: Fixed in 7811ceda — `upsertFrontMatterVerified` now locates the `doc-detective:` mapping first and only edits the `verified:` key nested within it, so an unrelated top-level `verified:` elsewhere in the front matter is never matched. Added a regression test (`front-matter scoping`).

File: test/verified-writeback.test.js [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+          '<!-- test {"testId": "wait-demo"} -->',
+          '<!-- step {"wait": 10} -->',
+          "<!-- test end -->",
+          "",
+          "<!-- verified id=wait-demo -->",
+          "",
+        ].join("\n")
+      );
+      await runTests({ input: f, logLevel: "silent" });
+      expect(read(f)).to.contain(`<!-- verified id=wait-demo date=${TODAY} -->`);
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply test/verified-writeback.test.js]: Fixed in 7811ceda — this E2E assertion now matches a `\d{4}-\d{2}-\d{2}` date pattern instead of a pre-`runTests` TODAY value, removing the UTC-rollover flake.

File: test/verified-writeback.test.js [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+          "",
+          '<!-- test {"testId": "cross-file-demo"} -->',
+          '<!-- step {"wait": 10} -->',
+          "<!-- test end -->",
+          "",
+        ].join("\n")
+      );
+      const prose = write("badge-page.md", "# Badge page\n\n<!-- verified id=cross-file-demo -->\n");
+      await runTests({ input: tmpDir, logLevel: "silent" });
+      expect(read(prose)).to.contain(`<!-- verified id=cross-file-demo date=${TODAY} -->`);
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply test/verified-writeback.test.js]: Fixed in 7811ceda — same pattern-based assertion applied here.

File: src/core/utils.ts [hunk 11]
```diff
@@ -1154,6 +1161,384 @@ function replaceEnvs(stringOrObject: any): any {
...
+      const imgLineRe = new RegExp(`^([ \\t]*\\r?\\n[ \\t]*)(?:${body})`);
+      const tail = content.slice(marker.end);
+      const mm = tail.match(imgLineRe);
+      if (mm) {
+        content =
+          content.slice(0, marker.end) + mm[1] + img + content.slice(marker.end + mm[0].length);
+      } else {
+        content = content.slice(0, marker.end) + eol + img + content.slice(marker.end);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts:1595]: Fixed in 2e0a688f — the insert branch now copies the marker line's leading whitespace, so a badge added to an indented marker (e.g. a nested list item) keeps the construct's indentation. Re-runs preserve it via the existing image-line regex (`mm[1]`). Added a test asserting the inserted image is indented and the result is idempotent.

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,18 @@
+# [4.17.0-last-verified-on.2](https://github.com/doc-detective/doc-detective/compare/v4.17.0-last-verified-on.1...v4.17.0-last-verified-on.2) (2026-06-30)
+
+
+### Bug Fixes
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply CHANGELOG.md:26]: This CHANGELOG section is auto-generated by semantic-release (via conventional-changelog) from the prerelease `chore(release): … [skip ci]` commits the release bot adds to `feat/**` branches — it is not hand-authored in this PR, and none of the feature commits touch `CHANGELOG.md`. Editing the heading here would be overwritten on the next release run, and the `### Bug Fixes` level comes from the changelog template (under the `##` version header). Leaving this to the release tooling; skipping as out of scope for this change.

File: adrs/01046-last-verified-on-marker-writeback.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+* Good: idempotent in-source writes; honest aging; author-explicit anchoring (no content-hash guessing,
+  no orphans); zero new config/schema surface; zero new CI code.
+* Good: implementation is contained in `src/core` (`applyVerifiedMarkers` + pure helpers in
+  `src/core/utils.ts`, wired into `runTests`), so the schemas and the report shape are untouched.
+* Coverage: the writer scans the **union** of the report's content paths and the full detected input
+  set, so a marker in a prose-only page that produces no spec — referencing a test/spec defined in
+  another file — is still updated. The detected input set is side-channelled from `qualifyFiles`
+  (`config._qualifiedFiles`) through `resolvedTests._qualifiedFiles` into the writer, reusing the exact
+  file discovery the run already performed (no re-enumeration, no duplicated globbing).
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply adrs/01046-last-verified-on-marker-writeback.md:77]: Fixed in 8901d876 — the ADR Consequences now documents the pre-resolved (`DOC_DETECTIVE_API`) exception: there `_qualifiedFiles` is unavailable, so the candidate set falls back to report content paths and prose-only pages referencing a test defined elsewhere are not scanned (a `debug` line notes it).

File: src/core/utils.ts [hunk 14]
```diff
@@ -1154,6 +1161,389 @@ function replaceEnvs(stringOrObject: any): any {
...
+  }
+}
+
+// Set `doc-detective.verified.date` in the leading front matter. The id is read
+// with the YAML parser (robust), but the write is a *surgical text edit* of only
+// the `date:` line within the block-style `verified:` mapping — never a re-stringify
+// of the whole document. Re-serializing reflows unrelated fields (e.g. an inline
+// `tags: [a, b]` gains inner padding), which would create noisy, unintended diffs.
+// Returns the original content unchanged if there is no front-matter verified block,
+// no date to write, or the block isn't the supported block-style shape.
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts]: Fixed in 8901d876 — flow-style `doc-detective: {verified: {id: …}}` no longer silently no-ops. When the block-style surgical edit can't locate the structure, the writer now falls back to a YAML re-serialize (`upsertFrontMatterViaYaml`) so the date is written on PASS, honoring the marker contract. Added a test covering the flow-style case.

File: src/core/utils.ts [hunk 15]
```diff
@@ -1154,6 +1161,389 @@ function replaceEnvs(stringOrObject: any): any {
...
+  const files = new Set<string>();
+  for (const file of detectedFiles) {
+    if (file) files.add(file);
+  }
+  for (const spec of results.specs) {
+    if (spec?.specId) resultById.set(spec.specId, spec.result);
+    if (spec?.contentPath) files.add(spec.contentPath);
+    for (const test of spec?.tests || []) {
+      if (test?.testId) resultById.set(test.testId, test.result);
+      if (test?.contentPath) files.add(test.contentPath);
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts]: Fixed in 8901d876 — the bulk path now resolves each id through `resolveVerifiedId`, which checks specId before testId, so a same-named id stamps against the spec result (consistent with the single-lookup helper). This also puts the previously test-only export on the real code path. Added a precedence regression test.

File: src/core/utils.ts [hunk 16]
```diff
@@ -1154,6 +1161,389 @@ function replaceEnvs(stringOrObject: any): any {
...
+    if (/^\s*date:/.test(lines[i])) dateIndex = i;
+  }
+  if (childIndent === null) childIndent = `${vIndent}  `;
+
+  if (dateIndex !== -1) {
+    lines[dateIndex] = lines[dateIndex].replace(/^(\s*date:\s*).*$/, `$1${date}`);
+  } else {
+    const insertAfter = idIndex !== -1 ? idIndex : lastChildIndex;
+    lines.splice(insertAfter + 1, 0, `${childIndent}date: ${date}`);
+  }
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts:1728]: Fixed in 8901d876 — the front-matter `date` is now written as a double-quoted string (both the surgical edit and the YAML fallback), so YAML 1.1 readers like js-yaml read `"YYYY-MM-DD"` as a string instead of coercing it to a Date/timestamp.

File: src/core/utils.ts [hunk 17]
```diff
@@ -1154,6 +1161,420 @@ function replaceEnvs(stringOrObject: any): any {
...
+    if (/^\s*id:/.test(lines[i])) idIndex = i;
+    if (/^\s*date:/.test(lines[i])) dateIndex = i;
+  }
+  if (childIndent === null) childIndent = `${vIndent}  `;
+
+  if (dateIndex !== -1) {
+    lines[dateIndex] = lines[dateIndex].replace(/^(\s*date:\s*).*$/, `$1"${date}"`);
+  } else {
+    const insertAfter = idIndex !== -1 ? idIndex : lastChildIndex;
+    lines.splice(insertAfter + 1, 0, `${childIndent}date: "${date}"`);
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts]: Fixed — `upsertFrontMatterVerified` now only records `id:`/`date:` lines at the direct-child indent of `verified:`, so a deeper `verified.meta.date` is no longer mistaken for `verified.date`. Added a regression test.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by hawkeyexl** [comment]
Rebased onto the latest `main` and squashed to a single feature commit to clear the merge conflict (the semantic-release version-bump commits were the conflict source). Also **renumbered the ADR to `01045`** — `01010` is now taken by the SSRF-guard ADR on main.

Addressed the latest review findings in this pass:
- **`verified.meta.date` mis-scope** (coderabbit): the front-matter scanner now only matches direct children of `verified:`, so a nested `meta.date` isn't mistaken for `verified.date`. Regression test added.
- **`.xml` → `dita` mapping** (claude nit): dropped — `.xml` is generic (POMs, SVGs, feeds); keeping `.dita`/`.ditamap` avoids putting unrelated files on the write-back scan path.
- **`setVerifiedAttrDate` quote normalization** (claude nit): documented as intentional (attribute is not YAML; parse strips quotes on read; idempotent thereafter).
- **Partial-pass coverage** (claude nit): added an integration test — two markers, different ids, only the passing one is stamped; the failing one ages.

Merged cleanly against main's `src/core` additions; verified suite is 54 passing locally.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/core/utils.ts [hunk 18]
```diff
@@ -1389,6 +1396,426 @@ function replaceEnvs(stringOrObject: any): any {
...
+    if (indentMatch[1] === childIndent) {
+      if (/^\s*id:/.test(lines[i])) idIndex = i;
+      if (/^\s*date:/.test(lines[i])) dateIndex = i;
+    }
+  }
+  if (childIndent === null) childIndent = `${vIndent}  `;
+
+  if (dateIndex !== -1) {
+    lines[dateIndex] = lines[dateIndex].replace(/^(\s*date:\s*).*$/, `$1"${date}"`);
+  } else {
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts:1725]: Fixed — the `date:` line rewrite now replaces only the value (`/^(\s*date:\s*)("[^"]*"|'[^']*'|[^\s#]*)/`), so a trailing inline comment (`date: "…" # note`) is preserved. Added a test.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/core/utils.ts [hunk 19]
```diff
@@ -1389,6 +1396,426 @@ function replaceEnvs(stringOrObject: any): any {
...
+  const eol = fm[0].includes("\r\n") ? "\r\n" : "\n";
+  const lines = fm[1].split(/\r?\n/);
+
+  // Locate the `doc-detective:` mapping first, then the `verified:` key nested
+  // within it — so an unrelated top-level `verified:` elsewhere in the front
+  // matter can't be matched by mistake.
+  let ddIndex = -1;
+  let ddIndent = "";
+  for (let i = 0; i < lines.length; i++) {
+    const m = lines[i].match(/^([ \t]*)(?:"doc-detective"|'doc-detective'|doc-detective):[ \t]*$/);
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts]: Fixed — the `doc-detective:` and `verified:` block-key detection now allows a trailing comment (`…:[ \t]*(?:#.*)?$`), so `doc-detective: # comment` still takes the surgical edit path instead of falling back to YAML re-serialization. Added a test.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/core/utils.ts [hunk 20]
```diff
@@ -1389,6 +1396,430 @@ function replaceEnvs(stringOrObject: any): any {
...
 
+// ===========================================================================
+// "Last Verified On" — marker-driven write-back.
+//
+// Authors place a static `verified` marker carrying an `id` (a test or spec id)
+// they choose; after a run we resolve that id to its roll-up result and, only
+// when it PASSED, write today's date back into the marker. On non-pass the date
+// is left untouched so the badge silently "ages". There is no config: the
+// feature acts wherever a marker exists, and nowhere else. See
+// adrs/01045-last-verified-on-marker-writeback.md.
```
[the remark this answers is not in the document]
Comment (hawkeyexl) [reply src/core/utils.ts]: Fixed — updated the PR description to reference `adrs/01045-last-verified-on-marker-writeback.md`. The ADR was renumbered from 01010 to 01045 during the rebase onto main, since 01010 is now taken by the SSRF-guard ADR; the code comment and the ADR file already use 01045.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
